tractor/_trionics.py

```diff
@@ -142,72 +145,123 @@ def __init__(self, actor: Actor) -> None:
 
     async def wait(self) -> None:
         """Wait for all subactors to complete.
+
+        This is probably the most complicated (and confusing, sorry)
+        function that does all the clever crap to deal with cancellation,
+        error propagation, and graceful subprocess tear down.
         """
-        async def maybe_consume_result(portal, actor):
-            if (
-                portal in self._cancel_after_result_on_exit and
-                (portal._result is None and portal._exc is None)
-            ):
-                log.debug(f"Waiting on final result from {subactor.uid}")
-                res = await portal.result()
-                # if it's an async-gen then we should alert the user
-                # that we're cancelling it
+        async def exhaust_portal(portal, actor):
+            """Pull final result from portal (assuming it was one).
+
+            If the main task is an async generator do our best to consume
+            what's left of it.
+            """
+            try:
+                log.debug(f"Waiting on final result from {actor.uid}")
+                final = res = await portal.result()
+                # if it's an async-gen then alert that we're cancelling it
                 if inspect.isasyncgen(res):
+                    final = []
                     log.warning(
                         f"Blindly consuming asyncgen for {actor.uid}")
                     with trio.fail_after(1):
                         async with aclosing(res) as agen:
                             async for item in agen:
                                 log.debug(f"Consuming item {item}")
+                                final.append(item)
+            except Exception as err:
+                # we reraise in the parent task via a ``trio.MultiError``
+                return err
+            else:
+                return final
+
+        async def cancel_on_completion(
+            portal: Portal,
+            actor: Actor,
+            task_status=trio.TASK_STATUS_IGNORED,
+        ) -> None:
+            """Cancel actor gracefully once it's "main" portal's
+            result arrives.
+
+            Should only be called for actors spawned with `run_in_actor()`.
+            """
+            with trio.open_cancel_scope() as cs:
+                task_status.started(cs)
+                # this may error in which case we expect the far end
```

Comment:
Actually this is also a wrong comment. We still cancel since `exhaust_portal()` doesn't raise any more (it used to)..